### PR TITLE
Task 3 unittests

### PR DESCRIPTION
**Pull request by oscamaina**

### CODE REVIEW

File: tests/test_database.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def setUp(self):
+        self.dojo = Dojo()
+
+    def test_successfully_saves_state(self):
+        """ Test saves application state to Database. """
+        self.dojo.create_room('living', ["kigali"])
+        self.dojo.create_room('office', ["Django"])
+        self.dojo.add_person("Maina", "oscar", "fellow")
+        output = self.dojo.save_state("somedb")
+        self.assertEqual("App data has been stored in somedb.db database", output)
```
Comment (Gchesta) [inline tests/test_database.py]: How does this test that the data has been saved successfully? You may have to query the database to see whether this has been done succssefully.

File: tests/test_database.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    def test_successfully_load_state(self):
+        """ Test loading dojo state from Database. """
+        initial_room_count= len(self.dojo.all_rooms)
+        self.dojo.load_state("somedb")
+        room_count_after = len(self.dojo.all_rooms)
+        self.assertEqual(room_count_after - initial_room_count, 5)
```
Comment (Gchesta) [inline tests/test_database.py]: Makes sense. Now check the length of the persons too, and their attributes.

File: tests/test_database.py (hunk 1 above)
Comment (Gchesta) [reply tests/test_database.py]: Another thing, the format for self.assert is (output, expected output). Just revise that.

Comment (oscamaina) [reply tests/test_database.py]: Thanks

File: tests/test_database.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    def test_successfully_saves_state(self):
+        """ Test saves application state to Database. """
+        self.dojo.create_room('living', ["kigali"])
+        self.dojo.create_room('office', ["Django"])
+        self.dojo.add_person("Maina", "oscar", "fellow")
+        output = self.dojo.save_state("flask.db")
+
+        #connect to db and create session
+        engine = create_engine('sqlite:///database/somedb.db')
```
Comment (mnjaggah) [inline tests/test_database.py]: you need to import session maker and create_engine

Comment (mnjaggah) [inline tests/test_database.py]: you have used different databases for this test, was that intentional

Comment (oscamaina) [reply tests/test_database.py]: It's imported in the dojo file, on this file i am importing everything from dojo

Comment (oscamaina) [reply tests/test_database.py]: typo error, fixed

File: tests/test_database.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        rooms = session.query(RoomModel).all()
+        persons = session.query(PersonModel).all()
+
+        self.assertEqual(len(rooms),2)
+        self.assertEqual(len(persons),2)
+        self.assertEqual(output,"App data has been stored in somedb.db database")
+
+    def test_successfully_load_state(self):
+        """ Test loading data from database to application """
+        self.dojo.load_state("flask.db")
```
Comment (ianoti) [inline tests/test_database.py]: why are you loading from different database from the one you saved data to?

File: tests/test_database.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    def test_successfully_saves_state(self):
+        """ Test saves application state to Database. """
+        self.dojo.create_room('living', ["kigali"])
+        self.dojo.create_room('office', ["Django"])
+        self.dojo.add_person("Maina", "oscar", "fellow")
+        self.dojo.add_person("Ken", "musomi", "staff")
+        output = self.dojo.save_state("somedb.db")
+
+        #connect to db and create session
+        engine = create_engine('sqlite:///database/somedb.db')
```
Comment (ianoti) [inline tests/test_database.py]: unittests should test functionality of the app itself, this `create_engine` section from lines 18-25 should not appear in your code. check the data exists by loading from the db and checking the loaded data is equivalent to what you expect.

**Review by ianoti — CHANGES_REQUESTED** [review]
A better way of testing the database functionality would be to have 2 instances of  Dojo(). in one instance of the Dojo() add people and add rooms and perform the save state functionality, in the second instance of Dojo use the load state functionality to recover the people and rooms in the first instance of Dojo. then you assert to check that both instances have equivalent rooms and persons hence testing save state and load state.

**Review by ianoti — CHANGES_REQUESTED** [review]
Your database tests are not logical. you are creating the rooms in the `fdojo` instance but saving the state of the `dojo` instance which has no rooms defined in it. it would be better to save state of `fdojo` and load state in `dojo` 
Another thing, improve the naming of the instances `fdojo` isn't descriptive enough.

**Review by jimmykimani — APPROVED** [review]

**Pull request closed without merging**